File: Temp/leetcode/editor/cn/[1486]数组异或操作.py
# 1486 数组异或操作
from functools import reduce


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def xorOperation(self, n: int, start: int) -> int:
        return reduce(lambda x, y: x ^ y, [start + i * 2 for i in range(n)])


# https://leetcode.cn/problems/xor-operation-in-an-array/solutions/2793723/o1-gong-shi-tui-dao-pythonjavaccgojsrust-le23
class Solution:
    def xorOperation(self, n: int, start: int) -> int:
        a = start // 2
        b = n & start & 1
        xumXOR = lambda n: (n, 1, n + 1, 0)[n % 4]
        return (xumXOR(a + n - 1) ^ xumXOR(a - 1)) * 2 + b

# leetcode submit region end(Prohibit modification and deletion)

File: Temp/leetcode/editor/cn/[1486]数组异或操作.c
// 1486 数组异或操作
// https://leetcode.cn/problems/xor-operation-in-an-array/


//leetcode submit region begin(Prohibit modification and deletion)
/*
int sumXOR(int n) {
    // 计算 0 到 n 的异或和
    switch (n % 4) {
    case 0:
        return n;
    case 1:
        return 1;
    case 2:
        return n + 1;
    default:
        return 0;
    }
}

int xorOperation(int n, int start) {
    int a = start / 2;
    int b = n & start & 1;
    return (sumXOR(a - 1) ^ sumXOR(a + n - 1)) * 2 + b;
}
*/

// https://leetcode.cn/problems/xor-operation-in-an-array/solutions/2793723/o1-gong-shi-tui-dao-pythonjavaccgojsrust-le23

int xorOperation(int n, int start) {
    int res = 0;
    for (int i = 0; i < n; i++) {
        res ^= start + 2 * i;
    }
    return res;
}
//leetcode submit region end(Prohibit modification and deletion)
